swarm/src/lib.rs

```diff
@@ -917,25 +966,42 @@ where
                 local_addr,
                 send_back_addr,
             } => {
-                let handler = self.behaviour.new_handler();
+                let connection_id = ConnectionId::default();
+
+                match self.behaviour.handle_pending_inbound_connection(
+                    connection_id,
+                    &local_addr,
+                    &send_back_addr,
+                ) {
+                    Ok(()) => {}
+                    Err(error) => {
+                        // TODO: self.behaviour.on_swarm_event(FromSwarm::ListenFailure())
+
+                        return Some(todo!("Report connection error event, these are currently modeled in a weird way where we split them by incoming/outgoing on the `SwarmEvent` level but everywhere else by established/pending"));
```

Comment:
@mxinden: Big question mark here on how we want to model this. Input welcome.

Reply:
If I understand correctly, we didn't return a `SwarmEvent` in this particular case at all, right? I am not opposed to returning a `SwarmEvent` in the case were we reject a pending inbound connection.

I don't have a strong opinion on which of the two namespace splits should take preference. Striving for consistency of the two sounds good to me, though seems to be a separate undertaking.

Reply:
> If I understand correctly, we didn't return a `SwarmEvent` in this particular case at all, right?

Right, we didn't. I hadn't considered _not_ returning one, that might indeed unblock this bit, thanks!

> Striving for consistency of the two sounds good to me, though seems to be a separate undertaking.

It is a bit of a mess 😔

Reply:
> > If I understand correctly, we didn't return a `SwarmEvent` in this particular case at all, right?
> 
> Right, we didn't. I hadn't considered _not_ returning one, that might indeed unblock this bit, thanks!
> 
> > Striving for consistency of the two sounds good to me, though seems to be a separate undertaking.
> 
> It is a bit of a mess pensive

Actually, this still proves to be a bit difficult. In follow-up PRs to this, I think we should re-implement some of the current functionality in regards to banning, connection limits etc on top of this new infrastructure. That will remove several, currently explicit error cases. I think if we don't model these error properly, it will be difficult for users to understand what is happening.

I am proposing to align these errors before we merge this PR.

Reply:
> I am proposing to align these errors before we merge this PR.

It might have to be part of this, I'll see what I can do.

Reply:
> I think we should re-implement some of the current functionality in regards to banning, connection limits etc on top of this new infrastructure.

Just to make sure we agree here, long term I would like all connection management to happen through `NetworkBehaviour` implementations. I.e. `libp2p-swarm` should not be aware of the concepts of connection limits nor peer banning.

Reply:
> I am proposing to align these errors before we merge this PR.

Fine with me.

Reply:
> > I think we should re-implement some of the current functionality in regards to banning, connection limits etc on top of this new infrastructure.
> 
> Just to make sure we agree here, long term I would like all connection management to happen through `NetworkBehaviour` implementations. I.e. `libp2p-swarm` should not be aware of the concepts of connection limits nor peer banning.

Yes. I assumed we would still provide a simple `Blocklist` behaviour that replaces the current banning functionality?